WebService/v2/Models/DeviceModelApiModel/DeviceModelApiModel.cs

```diff
@@ -0,0 +1,198 @@
+        }
+
+        // Map API model to service model
+        public DeviceModel ToServiceModel(string id = "")
+        {
+            this.Id = id;
+
+            var now = DateTimeOffset.UtcNow;
+
+            var result = new DeviceModel
+            {
+                ETag = this.ETag,
+                Id = this.Id,
+                Version = this.Version,
+                Name = this.Name,
+                Description = this.Description,
+                Type = this.Type,
+                Protocol = (IoTHubProtocol)Enum.Parse(typeof(IoTHubProtocol), this.Protocol, true),
+                Simulation = DeviceModelSimulation.ToServiceModel(this.Simulation),
+                Properties = new Dictionary<string, object>(this.Properties),
+                Telemetry = this.Telemetry.Select(x => DeviceModelTelemetry.ToServiceModel(x)).ToList(),
+                CloudToDeviceMethods = null
+            };
+
+            // Map the list of CloudToDeviceMethods
+            if (this.CloudToDeviceMethods != null && this.CloudToDeviceMethods.Count > 0)
+            {
+                result.CloudToDeviceMethods = new Dictionary<string, Script>();
+                foreach (KeyValuePair<string, DeviceModelSimulationScript> method in this.CloudToDeviceMethods)
+                {
+                    var fieldValue = method.Value.ToServiceModel();
+                    result.CloudToDeviceMethods.Add(method.Key, fieldValue);
+                }
+            }
+
+            return result;
+        }
+
+        // Map service model to API model
+        public static DeviceModelApiModel FromServiceModel(DeviceModel value)
+        {
+            if (value == null) return null;
+
+            var result = new DeviceModelApiModel
+            {
+                ETag = value.ETag,
+                Id = value.Id,
+                Version = value.Version,
+                Name = value.Name,
+                Description = value.Description,
+                created = value.Created,
+                modified =value.Modified,
+                Protocol = value.Protocol.ToString(),
+                Simulation = DeviceModelSimulation.FromServiceModel(value.Simulation)
+            };
+
+            foreach (var property in value.Properties)
+            {
+                result.Properties.Add(property.Key, property.Value);
+            }
+
+            foreach (var message in value.Telemetry)
+            {
+                result.Telemetry.Add(DeviceModelTelemetry.FromServiceModel(message));
+            }
+
+            if (value.CloudToDeviceMethods?.Count > 0)
+            {
+                foreach (var method in value.CloudToDeviceMethods)
+                {
+                    result.CloudToDeviceMethods.Add(method.Key, DeviceModelSimulationScript.FromServiceModel(method.Value));
+                }
+            }
+
+            return result;
+        }
+
+        public void ValidateInputRequest(ILogger log)
```

Comment:
add an extra param, e.g. `bool creating = true`. Use `true` with POST and `false` with PUT

Reply:
Done.